### PR TITLE
Add `code_actions` as `formatter` type

### LINKED ISSUE #8992: Add ability to only run `code_actions` on format, no formatter
### Check for existing issues

- [X] Completed

### Describe the feature

See: https://github.com/zed-industries/zed/issues/4325#issuecomment-1981161243

Essentially: users want to only format their code using ESLint (or other language server's code actions).

Problem: we have `code_actions_on_format`, which allow you to run code actions when formatting, but if you turn formatting off (`formatter: "off"`) then those won't get run either.

So we have a few options on how we want to enable this:

1. `"noop"`-formatter:

```json
{
  "TypeScript": {
    "formatter": "noop", // <-- new
    "code_actions_on_format": {
      "source.fixAll.eslint": true
    }
  }
}
```

2. `code_actions` as a formatter:

```json
{
  "TypeScript": {
    "formatter": {
      "code_actions": {
        "source.fixAll.eslint": true
      }
    }
  }
}
```

3. Introduce `code_actions_on_save` (save!):
```json
{
  "TypeScript": {
    "formatter": "off",
    "code_actions_on_save": {
      "source.fixAll.eslint": true
    }
  }
}
```

I'm not super happy with any of them, but might lean towards `noop` the most, since it's the simplest and doesn't introduce any new concepts and orderings. But it does feel weird to have a `noop` formatter and then run code actions? So from that perspective, (2) seems good. Problem with (2) is that it's unclear how it behaves when code actions on format are also set? Do we run it twice?

### If applicable, add mockups / screenshots to help present your vision of the feature

noop

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes #8992 and solves a problem that ESLint/Prettier/... users have been running into:

They want to format _only_ with ESLint, which is *not* a primary language server (so `formatter: language server` does not help) and it is not a formatter.

What they want to use is what they get when they have configured something like this:

```json
{
  "languages": {
    "JavaScript": {
      "code_actions_on_format": {
        "source.fixAll.eslint": true
      }
    }
  }
}
```

BUT they don't want to run the formatter.

So what this PR does is to add a new formatter type: `code_actions`.

With that, users can only use code actions to format:

```json
{
  "languages": {
    "JavaScript": {
      "formatter": {
        "code_actions": {
          "source.fixAll.eslint": true
        }
      }
    }
  }
}
```

This means that when formatting (via `editor: format` or on-save) only the code actions that are specified are being executed, no formatter.

Release Notes:

- Added a new `formatter`/`format_on_save` option: `code_actions`. When configured, this uses language server code actions to format a buffer. This can be used if one wants to, for example, format a buffer with ESLint and *not* run prettier or another formatter afterwards. Example configuration: `{"languages": {"JavaScript": {"formatter": {"code_actions": {"source.fixAll.eslint": true}}}}}` ([#8992](https://github.com/zed-industries/zed/issues/8992)).

### CODE REVIEW

File: crates/language/src/language_settings.rs [hunk 1]
```diff
@@ -292,6 +293,8 @@ pub enum FormatOnSave {
         /// The arguments to pass to the program.
         arguments: Arc<[String]>,
     },
+    // Files should be formatted using code actions executed by language servers.
```
Comment (maxdeviant) [inline crates/language/src/language_settings.rs:296]: This should be a doc comment:

```suggestion
    /// Files should be formatted using code actions executed by language servers.
```

File: crates/language/src/language_settings.rs [hunk 2]
```diff
@@ -325,6 +328,8 @@ pub enum Formatter {
...
```
Comment (maxdeviant) [inline crates/language/src/language_settings.rs:331]: Same here:

```suggestion
    /// Files should be formatted using code actions executed by language servers.
```

File: crates/project/src/project.rs [hunk 3]
```diff
@@ -4539,93 +4539,26 @@ impl Project {
...
+                    push_to_history,
+                    &mut project_transaction,
+                    &mut cx,
+                )
+                .await?;
             }
 
             // Apply language-specific formatting using either the primary language server
             // or external command.
+            // Except for cod
```
Comment (maxdeviant) [inline crates/project/src/project.rs:4561]: What does this comment mean?

Comment (mrnugget) [reply crates/project/src/project.rs:4562]: Sorry about that. Fixed now. Got lost in pairing session.

**Comment by zzzej** [comment]
This is different from the top-level `code_actions_on_format`?
```json
{
 "code_actions_on_format": {
    // "source.organizeImports": true,
    "source.fixAll.eslint": true
  },
}
```

**Comment by Mau-MD** [comment]
Love this! Tysm

**Comment by ramipellumbi** [comment]
This is awesome! I have been inching my way towards Zed from VsCode and I think this is what I needed to do my Typescript development on Zed full time (I almost exclusively use it for Python outside of Jupyter notebooks). Keep up the great work :)

**Pull request merged**